File: docs-examples/example-java/src/test/java/io/micronaut/jms/docs/binding/BindingSpec.java
package io.micronaut.jms.docs.binding;

import io.micronaut.jms.docs.AbstractJmsSpec;
import org.junit.jupiter.api.Test;

import java.io.Serializable;
import java.util.HashMap;
import java.util.Map;
import java.util.Random;
import java.util.UUID;

import static io.micronaut.jms.model.JMSHeaders.JMS_CORRELATION_ID;
import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;

class BindingSpec extends AbstractJmsSpec {

    @Test
    void testMapProducerAndConsumerWithHeaders() {
        int foo = 123;
        boolean bar = true;
        Map<String, Serializable> body = new HashMap<>();
        body.put("foo", foo);
        body.put("bar", bar);

        Random random = new Random();
        String correlationId = UUID.randomUUID().toString();
        String stringValue = UUID.randomUUID().toString();
        boolean booleanValue = random.nextBoolean();
        byte byteValue = (byte) random.nextInt(Byte.MAX_VALUE);
        short shortValue = (short) random.nextInt(Short.MAX_VALUE);
        int intValue = random.nextInt();
        long longValue = System.currentTimeMillis();
        float floatValue = random.nextFloat();
        double doubleValue = random.nextDouble();

        MapProducer mapProducer = applicationContext.getBean(MapProducer.class);
        mapProducer.send(
            body,
            correlationId,
            stringValue,
            booleanValue,
            byteValue,
            shortValue,
            intValue,
            longValue,
            floatValue,
            doubleValue);

        MapConsumer mapConsumer = applicationContext.getBean(MapConsumer.class);
        await().atMost(3, SECONDS).until(() ->
            mapConsumer.messageBodies.size() == 1 &&
                mapConsumer.messageBodies.get(0).get("foo").equals(foo) &&
                mapConsumer.messageBodies.get(0).get("bar").equals(bar) &&
                mapConsumer.messageHeaders.size() == 1 &&
                mapConsumer.messageHeaders.get(0).get(JMS_CORRELATION_ID).equals(correlationId) &&
                mapConsumer.messageHeaders.get(0).get("CustomStringHeader").equals(stringValue) &&
                mapConsumer.messageHeaders.get(0).get("CustomBooleanHeader").equals(booleanValue) &&
                mapConsumer.messageHeaders.get(0).get("CustomByteHeader").equals(byteValue) &&
                mapConsumer.messageHeaders.get(0).get("CustomShortHeader").equals(shortValue) &&
                mapConsumer.messageHeaders.get(0).get("CustomIntegerHeader").equals(intValue) &&
                mapConsumer.messageHeaders.get(0).get("CustomLongHeader").equals(longValue) &&
                mapConsumer.messageHeaders.get(0).get("CustomFloatHeader").equals(floatValue) &&
                mapConsumer.messageHeaders.get(0).get("CustomDoubleHeader").equals(doubleValue)
        );
    }
}

File: docs-examples/example-kotlin/src/test/kotlin/io/micronaut/jms/docs/binding/BindingSpec.kt
package io.micronaut.jms.docs.binding

import io.micronaut.inject.qualifiers.Qualifiers
import io.micronaut.jms.docs.AbstractJmsSpec
import jakarta.jms.ConnectionFactory
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

import io.micronaut.jms.activemq.classic.configuration.ActiveMqClassicConfiguration.CONNECTION_FACTORY_BEAN_NAME
import io.micronaut.jms.model.JMSHeaders.JMS_CORRELATION_ID
import org.awaitility.Awaitility.await
import java.util.Random
import java.util.UUID
import java.util.concurrent.TimeUnit

class BindingSpec : AbstractJmsSpec() {

    @Test
    fun testMapProducerAndConsumerWithHeaders() {
        assertTrue(applicationContext.containsBean(ConnectionFactory::class.java))
        assertTrue(applicationContext.containsBean(ConnectionFactory::class.java, Qualifiers.byName(CONNECTION_FACTORY_BEAN_NAME)))

        val foo = 123
        val bar = true
        val body = mapOf<String, java.io.Serializable>("foo" to foo, "bar" to bar)

        val random = Random()
        val correlationId = UUID.randomUUID().toString()
        val stringValue = UUID.randomUUID().toString()
        val booleanValue = random.nextBoolean()
        val byteValue = random.nextInt(Byte.MAX_VALUE.toInt()).toByte()
        val shortValue = random.nextInt(Short.MAX_VALUE.toInt()).toShort()
        val intValue = random.nextInt()
        val longValue = System.currentTimeMillis()
        val floatValue = random.nextFloat()
        val doubleValue = random.nextDouble()

        val mapProducer = applicationContext.getBean(MapProducer::class.java)
        mapProducer.send(
            body,
            correlationId,
            stringValue,
            booleanValue,
            byteValue,
            shortValue,
            intValue,
            longValue,
            floatValue,
            doubleValue
        )

        val mapConsumer = applicationContext.getBean(MapConsumer::class.java)
        await().atMost(3, TimeUnit.SECONDS).until {
            mapConsumer.messageBodies.size == 1 &&
                mapConsumer.messageBodies[0]["foo"] == foo &&
                mapConsumer.messageBodies[0]["bar"] == bar &&
                mapConsumer.messageHeaders.size == 1 &&
                mapConsumer.messageHeaders[0][JMS_CORRELATION_ID] == correlationId &&
                mapConsumer.messageHeaders[0]["CustomStringHeader"] == stringValue &&
                mapConsumer.messageHeaders[0]["CustomBooleanHeader"] == booleanValue &&
                mapConsumer.messageHeaders[0]["CustomByteHeader"] == byteValue &&
                mapConsumer.messageHeaders[0]["CustomShortHeader"] == shortValue &&
                mapConsumer.messageHeaders[0]["CustomIntegerHeader"] == intValue &&
                mapConsumer.messageHeaders[0]["CustomLongHeader"] == longValue &&
                mapConsumer.messageHeaders[0]["CustomFloatHeader"] == floatValue &&
                mapConsumer.messageHeaders[0]["CustomDoubleHeader"] == doubleValue
        }
    }
}
